File: sdk/typescript/src/namespaces/introspection.ts
/**
 * Introspection Namespace API
 *
 * Provides access to system introspection and agent information.
 * Useful for debugging, monitoring, and understanding system state.
 */

/**
 * Agent information from introspection
 */
export interface AgentInfo {
  name: string;
  type: string;
  provider: string;
  model: string;
  available: boolean;
  capabilities: string[];
  rate_limits?: {
    requests_per_minute: number;
    tokens_per_minute: number;
  };
  metadata?: Record<string, unknown>;
}

/**
 * Agent leaderboard entry
 */
export interface LeaderboardEntry {
  agent: string;
  elo_rating: number;
  wins: number;
  losses: number;
  draws: number;
  total_debates: number;
  win_rate: number;
  rank: number;
}

/**
 * Full system introspection
 */
export interface SystemIntrospection {
  version: string;
  agents: AgentInfo[];
  features: Record<string, boolean>;
  capabilities: string[];
  limits: Record<string, number>;
  uptime_seconds: number;
}

/**
 * Internal client interface
 */
interface IntrospectionClientInterface {
  get<T>(path: string): Promise<T>;
}

/**
 * Introspection API namespace.
 *
 * Provides methods for system introspection:
 * - Get full system state
 * - View agent leaderboards
 * - List available agents
 * - Get specific agent details
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai', apiKey: 'your-key' });
 *
 * // Get full system introspection
 * const system = await client.introspection.getAll();
 *
 * // Get agent leaderboard
 * const leaderboard = await client.introspection.getLeaderboard();
 *
 * // List available agents
 * const agents = await client.introspection.listAgents();
 * ```
 */
export class IntrospectionAPI {
  constructor(private client: IntrospectionClientInterface) {}

  /**
   * Get full system introspection.
   */
  async getAll(): Promise<SystemIntrospection> {
    return this.client.get('/api/v1/introspection/all');
  }

  /**
   * Get agent leaderboard.
   */
  async getLeaderboard(options?: { limit?: number }): Promise<{ leaderboard: LeaderboardEntry[] }> {
    const params = options?.limit ? `?limit=${options.limit}` : '';
    return this.client.get(`/api/v1/introspection/leaderboard${params}`);
  }

  /**
   * List available agents.
   */
  async listAgents(): Promise<{ agents: AgentInfo[] }> {
    return this.client.get('/api/v1/introspection/agents');
  }

  /**
   * Get specific agent details.
   */
  async getAgent(name: string): Promise<AgentInfo> {
    return this.client.get(`/api/v1/introspection/agents/${name}`);
  }

  /**
   * Check agent availability.
   */
  async checkAvailability(): Promise<{ available: string[]; unavailable: string[] }> {
    return this.client.get('/api/v1/introspection/agents/availability');
  }
}

File: sdk/python/aragora/namespaces/introspection.py
"""
Introspection Namespace API

Provides access to system introspection and agent information.
Useful for debugging, monitoring, and understanding system state.

Features:
- Get full system introspection
- View agent leaderboards
- List available agents
- Check agent availability
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class IntrospectionAPI:
    """
    Synchronous Introspection API.

    Provides methods for system introspection:
    - Get full system state
    - View agent leaderboards
    - List available agents
    - Get specific agent details

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai", api_key="...")
        >>> system = client.introspection.get_all()
        >>> agents = client.introspection.list_agents()
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    def get_all(self) -> dict[str, Any]:
        """
        Get full system introspection.

        Returns:
            Dict with:
            - version: System version
            - agents: List of agent info
            - features: Feature flags
            - capabilities: System capabilities
            - limits: System limits
            - uptime_seconds: Server uptime
        """
        return self._client.request("GET", "/api/v1/introspection/all")

    def get_leaderboard(self, limit: int | None = None) -> dict[str, Any]:
        """
        Get agent leaderboard.

        Args:
            limit: Maximum number of entries

        Returns:
            Dict with leaderboard entries containing:
            - agent: Agent name
            - elo_rating: ELO score
            - wins/losses/draws: Record
            - total_debates: Total debates
            - win_rate: Win percentage
            - rank: Current rank
        """
        params = {"limit": limit} if limit else None
        return self._client.request("GET", "/api/v1/introspection/leaderboard", params=params)

    def list_agents(self) -> dict[str, Any]:
        """
        List available agents.

        Returns:
            Dict with agents list containing:
            - name: Agent name
            - type: Agent type
            - provider: Model provider
            - model: Model name
            - available: Availability status
            - capabilities: Agent capabilities
        """
        return self._client.request("GET", "/api/v1/introspection/agents")

    def get_agent(self, name: str) -> dict[str, Any]:
        """
        Get specific agent details.

        Args:
            name: Agent name

        Returns:
            Dict with full agent information
        """
        return self._client.request("GET", f"/api/v1/introspection/agents/{name}")

    def check_availability(self) -> dict[str, Any]:
        """
        Check agent availability.

        Returns:
            Dict with:
            - available: List of available agents
            - unavailable: List of unavailable agents
        """
        return self._client.request("GET", "/api/v1/introspection/agents/availability")


class AsyncIntrospectionAPI:
    """
    Asynchronous Introspection API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     system = await client.introspection.get_all()
        ...     agents = await client.introspection.list_agents()
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def get_all(self) -> dict[str, Any]:
        """Get full system introspection."""
        return await self._client.request("GET", "/api/v1/introspection/all")

    async def get_leaderboard(self, limit: int | None = None) -> dict[str, Any]:
        """Get agent leaderboard."""
        params = {"limit": limit} if limit else None
        return await self._client.request("GET", "/api/v1/introspection/leaderboard", params=params)

    async def list_agents(self) -> dict[str, Any]:
        """List available agents."""
        return await self._client.request("GET", "/api/v1/introspection/agents")

    async def get_agent(self, name: str) -> dict[str, Any]:
        """Get specific agent details."""
        return await self._client.request("GET", f"/api/v1/introspection/agents/{name}")

    async def check_availability(self) -> dict[str, Any]:
        """Check agent availability."""
        return await self._client.request("GET", "/api/v1/introspection/agents/availability")
